### ISSUE TITLE
Cannot build atomics example with LTO enabled

### ISSUE THREAD

**Issue by gyrovorbis** [KallistiOS/KallistiOS#367]
...and I really cannot explain why, and no matter what I try, I cannot get it to link properly. 

The example located at `kos/examples/dreamcast/basic/threading/atomics` fails to link with LTO enabled. This is a known issue:
```
/opt/toolchains/dc/sh-elf-13.2.0/bin/../lib/gcc/sh-elf/13.2.0/../../../../sh-elf/bin/ld: /tmp/cc173KsQ.ltrans0.ltrans.o: in function `_atomic_add_buffer':
/opt/toolchains/dc/kos/examples/dreamcast/basic/threading/atomics/atomics.c:109:(.text.atomic_add_buffer+0xd4): undefined reference to `___atomic_load'
/opt/toolchains/dc/sh-elf-13.2.0/bin/../lib/gcc/sh-elf/13.2.0/../../../../sh-elf/bin/ld: /opt/toolchains/dc/kos/examples/dreamcast/basic/threading/atomics/atomics.c:109:(.text.atomic_add_buffer+0xe8): undefined reference to `___atomic_compare_exchange'
/opt/toolchains/dc/sh-elf-13.2.0/bin/../lib/gcc/sh-elf/13.2.0/../../../../sh-elf/bin/ld: /tmp/cc173KsQ.ltrans0.ltrans.o: in function `_thread.lto_priv.0':
/opt/toolchains/dc/kos/examples/dreamcast/basic/threading/atomics/atomics.c:121:(.text.thread.lto_priv.0+0x230): undefined reference to `___atomic_fetch_add_8'
/opt/toolchains/dc/sh-elf-13.2.0/bin/../lib/gcc/sh-elf/13.2.0/../../../../sh-elf/bin/ld: /tmp/cc173KsQ.ltrans1.ltrans.o: in function `main.constprop.0':
/opt/toolchains/dc/kos/examples/dreamcast/basic/threading/atomics/atomics.c:266:(.text.main.constprop.0+0x158): undefined reference to `___atomic_is_lock_free'
/opt/toolchains/dc/sh-elf-13.2.0/bin/../lib/gcc/sh-elf/13.2.0/../../../../sh-elf/bin/ld: /opt/toolchains/dc/kos/examples/dreamcast/basic/threading/atomics/atomics.c:325:(.text.main.constprop.0+0x400): undefined reference to `___atomic_load_8'
/opt/toolchains/dc/sh-elf-13.2.0/bin/../lib/gcc/sh-elf/13.2.0/../../../../sh-elf/bin/ld: /opt/toolchains/dc/kos/examples/dreamcast/basic/threading/atomics/atomics.c:325:(.text.main.constprop.0+0x414): undefined reference to `___atomic_load'
collect2: error: ld returned 1 exit status
make: *** [Makefile:22: atomics.elf] Error 1
```

If anyone has any ideas, I'm all ears. This one has stumped me. I'm worried there's some fundamental incompatibility between LTO and the pseudo compiler-intrinsics that resolve to KOS's atomics back-end at link-time... No idea.

**Comment by ijacquez** [comment]
I think this calls for another case of `__attribute__ ((used))`? I would add it to the `ATOMIC_*` macros as well as `__atomic_is_lock_free` and test if that works.

**Comment by gyrovorbis** [comment]
> I think this calls for another case of `__attribute__ ((used))`? I would add it to the `ATOMIC_*` macros as well as `__atomic_is_lock_free` and test if that works.

That was actually the first thing I tried, and no matter what I did or where I peppered that attribute on the these function definitions, the linker error still persisted and they were still removed by LTO... Honestly really confused. lol.

**Comment by ijacquez** [comment]
Here's another idea...

I would try designating a section (`.gcc_builtins`) and specify that these functions live in that section (`__attribute__ ((section(".gcc_builtins"))`).

From that, in the main linkerscript, write something like:

```
KEEP (*(.gcc_builtins)
KEEP (*(.gcc_builtins.*)
```

I think you could put that in the `.text` section.

**Comment by gyrovorbis** [comment]
OH SHIT. I'll give that one a try!

**Comment by andressbarajas** [comment]
Im not sure if I am testing this right. I edited my environ.sh by uncommenting this line:

export KOS_CFLAGS="${KOS_CFLAGS} -freorder-blocks-algorithm=simple -flto=auto"

I rebuilt KOS and then was able to compile the examples/dreamcast/basic/threading/atomics example without issue. @gyrovorbis Care to double check when you get the chance?

**Comment by darcagn** [comment]
With the default KOS `environ.sh` settings, with one change:
```
#export KOS_CFLAGS="-O2 -fomit-frame-pointer"
export KOS_CFLAGS="-O3 -flto=auto -fomit-frame-pointer"
```
I see the same error in the atomics demo as Falco. I tried `-O2` and tried adding `-freorder-blocks-algorithm=simple` to closer match your settings and I still have the same issue with those as well.

**Comment by gyrovorbis** [comment]
So everything tried so far has unfortunately not yielded any results... It really seems as though the configurable atomics model stuff going on for the SH architecture causes those symbols to be treated differently at the linking stage from regular functions, which causes them to be potentially incompatible with LTO...

Before I open an issue on the GCC Bugzilla, I had one last idea that could be tried. I'm writing it here so I don't forget, because I might not be able to get around to it for awhile.

I wonder if we can patch our custom atomics implementation into GCC itself as part of the "KOS-threads" model when we build the toolchain? I wonder if that would prevent LTO from removing the symbols at link-time? Not sure, but it's definitely worth a shot sometime if I or someone else gets a chance.

**Comment by pcercuei** [comment]
Fixed in #529.
